File: apple/RecKit/Sources/RecKit/Design/LedgerRow.swift
import SwiftUI

/// docs/09 화면 원칙 2: the ledger's four columns, in the widths the mockup sets.
private enum Column {
    static let time: CGFloat = 62
    static let length: CGFloat = 44
    /// Wide enough for `NEEDS_AUTH`, the longest code `LedgerStatus` mints.
    static let status: CGFloat = 96
}

/// The column headings above the ledger. At an accessibility type size the row below stops being
/// four columns (see [LedgerRow]), so there is nothing left for these to head.
public struct LedgerHeader: View {
    @Environment(\.blueprint) private var blueprint
    @Environment(\.dynamicTypeSize) private var typeSize
    private let time: String
    private let title: String
    private let length: String
    private let status: String

    public init(time: String, title: String, length: String, status: String) {
        self.time = time
        self.title = title
        self.length = length
        self.status = status
    }

    public var body: some View {
        VStack(spacing: 0) {
            if !typeSize.isAccessibilitySize {
                columns
            }
            HairLine()
        }
    }

    private var columns: some View {
        VStack(spacing: 0) {
            HStack(spacing: 10) {
                heading(time).frame(width: Column.time, alignment: .leading)
                heading(title).frame(maxWidth: .infinity, alignment: .leading)
                heading(length).frame(width: Column.length, alignment: .trailing)
                heading(status).frame(width: Column.status, alignment: .center)
            }
            .padding(.horizontal, Space.m)
            .padding(.vertical, 6)
            .accessibilityHidden(true)
        }
    }

    private func heading(_ text: String) -> some View {
        Text(verbatim: text)
            .font(blueprint.fonts.label)
            .tracking(0.6)
            .foregroundStyle(blueprint.palette.textMuted)
            .lineLimit(1)
    }
}

/// One recording, as a row of the ledger: when (monospace), what, how long, and the state as a code
/// (docs/09 화면 원칙 2). The whole row is one accessibility element — four separate announcements of
/// `08-29`, a title, `42:10` and `DONE` are worse than one sentence — so the caller hands in the
/// sentence, with the state in words rather than as a code.
///
/// A row that opens under itself says so: whether it is open is the element's value, and the tap is
/// offered a second time as a named action, so a screen reader can tell an open row from a closed
/// one without tapping it to find out (docs/09 "접근성", as Android's `LedgerRow` does with
/// `expand`/`collapse`).
public struct LedgerRow<Trailing: View>: View {
    @Environment(\.blueprint) private var blueprint
    @Environment(\.dynamicTypeSize) private var typeSize
    /// docs/07 rule 3: the row's own two sentences are resolved outside SwiftUI, and reading the
    /// locale is what declares the dependency that redraws them when the language changes.
    @Environment(\.locale) private var locale
    private let date: String
    private let time: String
    private let title: String
    private let subtitle: String
    private let length: String
    private let status: LedgerStatus
    private let announce: String
    private let expanded: Bool?
    private let action: () -> Void
    private let trailing: Trailing

    /// - Parameter expanded: whether the expansion under this row is open, on the shells whose row
    ///   opens one — nil where the tap does something else entirely (the Mac's window opens a
    ///   detail), because a row that announces an expand action nobody can see is worse than one
    ///   that says nothing.
    /// - Parameter trailing: what the row offers *about* the recording, beside the state and on the
    ///   same line — a chip, not a button's worth of height. It is a sibling of the row's own
    ///   button and never inside it, because a button with a button in it is one target that
    ///   swallows the other; it is laid out at its own width, so it takes a column rather than a
    ///   share of the row.
    public init(
        date: String,
        time: String,
        title: String,
        subtitle: String,
        length: String,
        status: LedgerStatus,
        announce: String,
        expanded: Bool? = nil,
        action: @escaping () -> Void,
        @ViewBuilder trailing: () -> Trailing
    ) {
        self.date = date
        self.time = time
        self.title = title
        self.subtitle = subtitle
        self.length = length
        self.status = status
        self.announce = announce
        self.expanded = expanded
        self.action = action
        self.trailing = trailing()
    }

    public var body: some View {
        VStack(spacing: 0) {
            // Centred, so the chip lands on the line the badge is on — which is the middle of the
            // row, whether the row is two lines of columns or the stack an accessibility size gets.
            HStack(spacing: 0) {
                if let expanded {
                    button
                        .accessibilityValue(Text(verbatim: openState(expanded)))
                        .accessibilityAction(named: Text(verbatim: toggleLabel(expanded))) {
                            action()
                        }
                } else {
                    button
                }
                trailing
                    .fixedSize()
                    .padding(.trailing, Space.m)
            }
            HairLine()
        }
        .background(blueprint.palette.surface)
    }

    /// The row itself — a real button rather than a tap gesture, so it is announced as something to
    /// press and can be reached from the keyboard.
    private var button: some View {
        Button(action: action) {
            Group {
                if typeSize.isAccessibilitySize { stacked } else { columns }
            }
            .lineLimit(1)
            .padding(.horizontal, Space.m)
            .padding(.vertical, 12)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(Text(verbatim: announce))
        .accessibilityAddTraits(.isButton)
    }

    /// What the tap would do, named — the same two sentences Android's row labels its
    /// `expand`/`collapse` actions with.
    private func toggleLabel(_ expanded: Bool) -> String {
        expanded
            ? RecKitStrings.localized("Collapse")
            : RecKitStrings.localized("Expand")
    }

    /// The row's value: what a screen reader says after the sentence, so an open row and a closed
    /// one are not announced identically.
    private func openState(_ expanded: Bool) -> String {
        expanded ? RecKitStrings.localized("Expanded") : RecKitStrings.localized("Collapsed")
    }

    /// Four columns, as docs/09 화면 원칙 2 draws the ledger.
    private var columns: some View {
        HStack(spacing: 10) {
            when.frame(width: Column.time, alignment: .leading)
            what.frame(maxWidth: .infinity, alignment: .leading)
            howLong
                .frame(width: Column.length, alignment: .trailing)
            StatusBadge(status).frame(width: Column.status)
        }
    }

    /// The same four things at an accessibility type size, where four columns across a phone is
    /// four truncations: the row keeps its order and turns ninety degrees.
    private var stacked: some View {
        VStack(alignment: .leading, spacing: 6) {
            what
            HStack(spacing: 12) {
                when
                howLong
                Spacer(minLength: 0)
            }
            // On a line of its own: the code is the state (docs/09 화면 원칙 2), and at these sizes
            // `NEEDS_AUTH` is as wide as the stamp and the length together.
            StatusBadge(status)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    private var when: some View {
        VStack(alignment: .leading, spacing: 1) {
            Text(verbatim: date)
            Text(verbatim: time)
        }
        .font(blueprint.fonts.monoSmall)
        .foregroundStyle(blueprint.palette.textMuted)
    }

    private var what: some View {
        VStack(alignment: .leading, spacing: 3) {
            Text(verbatim: title)
                .font(blueprint.fonts.rowTitle)
                .foregroundStyle(blueprint.palette.text)
            Text(verbatim: subtitle)
                .font(blueprint.fonts.monoSmall)
                .foregroundStyle(blueprint.palette.textMuted)
        }
    }

    private var howLong: some View {
        Text(verbatim: length)
            .font(blueprint.fonts.monoSmall)
            .foregroundStyle(blueprint.palette.text)
    }
}

extension LedgerRow where Trailing == EmptyView {
    /// The ledger as three of the four surfaces draw it: the row and nothing beside it.
    public init(
        date: String,
        time: String,
        title: String,
        subtitle: String,
        length: String,
        status: LedgerStatus,
        announce: String,
        expanded: Bool? = nil,
        action: @escaping () -> Void
    ) {
        self.init(
            date: date,
            time: time,
            title: title,
            subtitle: subtitle,
            length: length,
            status: status,
            announce: announce,
            expanded: expanded,
            action: action
        ) { EmptyView() }
    }
}

File: android/app/src/main/kotlin/app/recly/android/ui/component/LedgerRow.kt
package app.recly.android.ui.component

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.collapse
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.expand
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import app.recly.android.ui.theme.Space
import app.recly.android.ui.theme.blueprint
import app.recly.android.ui.theme.mono

/** docs/09 화면 원칙 2: the ledger's date column, in the width the mockup sets. */
private val TIME_COLUMN = 62.dp

/** The gap between two columns of the ledger. */
private val COLUMN_GAP = Space.s

/**
 * The two columns of the ledger that hold data rather than prose, and so are measured rather than
 * fixed. Both used to be a constant and both lost letters at a font scale of 1.3: `NEEDS_AUTH` in a
 * 76dp column, and `00:09` in a 44dp one, which came out as `00:0`.
 */
data class LedgerColumns(val length: Dp, val status: Dp)

/**
 * The narrowest a title column may be and still be a title: below this it holds a word and an
 * ellipsis, which is not what the ledger is for.
 */
val LedgerTitleMin: Dp = 96.dp

/** What shape the ledger takes at the width it was given. */
enum class LedgerLayout {
    /** when · what · how long · state, on one line. */
    COLUMNS,

    /** when · what, then how long and the state under them. */
    STACKED,
}

/**
 * Four columns, or two lines? The two data columns are measured (see [ledgerColumns]), so on a
 * narrow screen at a large font size they and the fixed time column can take everything: at 320dp
 * and a font scale of 1.3 the title was left some 34dp, and at larger scales nothing at all — at
 * which point the measured columns are squeezed too and the clipping this design measured its way
 * out of comes back. Whenever what is left for the title falls under [titleMin] the row gives up
 * being a table and becomes two lines instead, where the title has the whole width.
 */
fun ledgerLayout(
    available: Dp,
    columns: LedgerColumns,
    titleMin: Dp = LedgerTitleMin,
): LedgerLayout {
    // Everything the title does not get: the row's own padding, the time column, the two measured
    // columns, and the three gaps between the four of them.
    val taken = Space.m * 2 + TIME_COLUMN + columns.length + columns.status + COLUMN_GAP * 3
    return if (available - taken < titleMin) LedgerLayout.STACKED else LedgerLayout.COLUMNS
}

/**
 * [LedgerColumns] for the [lengths] and status [codes] this ledger can show, measured in the style
 * they are drawn in — so the widths carry the fluid scale and the user's own font size, and neither
 * column changes from row to row.
 */
@Composable
fun ledgerColumns(lengths: List<String>, codes: List<String>): LedgerColumns = LedgerColumns(
    length = textColumnWidth(lengths, mono.small),
    status = statusColumnWidth(codes),
)

/**
 * Where the title column starts, so a detail drawn under a row — what an expanded row says about
 * itself — lines up with the title rather than with an indent somebody guessed.
 */
val LedgerTitleInset: Dp = Space.m + TIME_COLUMN + COLUMN_GAP

/** A 1dp rule — the only divider this design has. */
@Composable
fun HairLine(modifier: Modifier = Modifier) {
    val palette = blueprint
    Box(
        modifier
            .fillMaxWidth()
            .height(palette.line)
            .background(palette.grid),
    )
}

/**
 * The column headings above the ledger. A stacked ledger ([LedgerLayout.STACKED]) heads two lines
 * rather than four columns, in the same widths its rows use, so each heading still sits over what
 * it names.
 */
@Composable
fun LedgerHeader(
    time: String,
    title: String,
    length: String,
    status: String,
    columns: LedgerColumns,
    modifier: Modifier = Modifier,
    layout: LedgerLayout = LedgerLayout.COLUMNS,
) {
    Column(modifier) {
        val head = Modifier
            .fillMaxWidth()
            .padding(horizontal = Space.m, vertical = Space.s)
            .clearAndSetSemantics {}
        when (layout) {
            LedgerLayout.COLUMNS -> Row(head, Arrangement.spacedBy(COLUMN_GAP)) {
                Heading(time, Modifier.width(TIME_COLUMN))
                Heading(title, Modifier.weight(1f))
                Heading(length, Modifier.width(columns.length), TextAlign.End)
                Heading(status, Modifier.width(columns.status), TextAlign.Center)
            }

            LedgerLayout.STACKED -> Column(head, Arrangement.spacedBy(Space.xs)) {
                Row(Modifier.fillMaxWidth(), Arrangement.spacedBy(COLUMN_GAP)) {
                    Heading(time, Modifier.width(TIME_COLUMN))
                    Heading(title, Modifier.weight(1f))
                }
                Row(Modifier.fillMaxWidth(), Arrangement.spacedBy(COLUMN_GAP, Alignment.End)) {
                    Heading(length, Modifier.width(columns.length), TextAlign.End)
                    Heading(status, Modifier.width(columns.status), TextAlign.Center)
                }
            }
        }
        HairLine()
    }
}

@Composable
private fun Heading(text: String, modifier: Modifier = Modifier, align: TextAlign = TextAlign.Start) {
    Text(
        text,
        modifier = modifier,
        style = MaterialTheme.typography.labelSmall,
        color = blueprint.textMuted,
        textAlign = align,
        maxLines = 1,
    )
}

/**
 * One recording, as a row of the ledger: when (monospace), what, how long, and the state as a code
 * (docs/09 화면 원칙 2). The whole row is one accessibility node — four separate announcements of
 * `08-29`, a title, `42:10` and `DONE` are worse than one sentence — so the caller hands in the
 * sentence, with the state in words rather than as a code.
 *
 * The row also opens a block under itself, and that is a fact a screen reader can otherwise only
 * find out by tapping: [expanded] becomes the row's `expand`/`collapse` action, labelled by
 * [toggleLabel] (docs/09 "접근성").
 */
@Composable
fun LedgerRow(
    date: String,
    time: String,
    title: String,
    subtitle: String,
    length: String,
    status: LedgerStatus,
    columns: LedgerColumns,
    announce: String,
    expanded: Boolean,
    toggleLabel: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    layout: LedgerLayout = LedgerLayout.COLUMNS,
) {
    val palette = blueprint
    Column(modifier.background(palette.surface)) {
        // One node and one tap, whichever shape the row takes: the sentence and the expand action
        // are on the body, and the two layouts below only rearrange what it wraps.
        val body = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .semantics(mergeDescendants = true) {
                contentDescription = announce
                role = Role.Button
                if (expanded) {
                    collapse(label = toggleLabel) { onClick(); true }
                } else {
                    expand(label = toggleLabel) { onClick(); true }
                }
            }
            .padding(horizontal = Space.m, vertical = 12.dp)
        when (layout) {
            LedgerLayout.COLUMNS -> Row(
                modifier = body,
                horizontalArrangement = Arrangement.spacedBy(COLUMN_GAP),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                TimeCell(date, time)
                TitleCell(title, subtitle, Modifier.weight(1f))
                LengthCell(length, Modifier.width(columns.length))
                StatusCell(status, Modifier.width(columns.status))
            }

            // The title has the width to itself; how long and what state go under it, right-aligned
            // in the widths the header uses so the two lines still read as a table.
            LedgerLayout.STACKED -> Column(body, Arrangement.spacedBy(Space.xs)) {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.spacedBy(COLUMN_GAP),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    TimeCell(date, time)
                    TitleCell(title, subtitle, Modifier.weight(1f))
                }
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.spacedBy(COLUMN_GAP, Alignment.End),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    LengthCell(length, Modifier.width(columns.length))
                    StatusCell(status, Modifier.width(columns.status))
                }
            }
        }
        HairLine()
    }
}

@Composable
private fun TimeCell(date: String, time: String) {
    val palette = blueprint
    Column(Modifier.width(TIME_COLUMN).clearAndSetSemantics {}) {
        Text(date, style = mono.small, color = palette.textMuted, maxLines = 1)
        Text(time, style = mono.small, color = palette.textMuted, maxLines = 1)
    }
}

@Composable
private fun TitleCell(title: String, subtitle: String, modifier: Modifier = Modifier) {
    val palette = blueprint
    Column(
        modifier = modifier.clearAndSetSemantics {},
        verticalArrangement = Arrangement.spacedBy(Space.xs),
    ) {
        Text(
            title,
            style = MaterialTheme.typography.titleSmall,
            color = palette.text,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
        )
        Text(
            subtitle,
            style = mono.small,
            color = palette.textMuted,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
        )
    }
}

@Composable
private fun LengthCell(length: String, modifier: Modifier = Modifier) {
    Text(
        length,
        modifier = modifier.clearAndSetSemantics {},
        style = mono.small,
        color = blueprint.text,
        textAlign = TextAlign.End,
        maxLines = 1,
    )
}

@Composable
private fun StatusCell(status: LedgerStatus, modifier: Modifier = Modifier) {
    Box(modifier.clearAndSetSemantics {}, contentAlignment = Alignment.Center) {
        StatusBadge(status)
    }
}
